src/backend/distributed/shardsplit/shardsplit_logical_replication.c

```diff
@@ -0,0 +1,734 @@
+ * shardSplitSubscriberMetadataList - List of subscriber metadata.
+ *
+ * sourceWorkerNode - Source node.
+ */
+void
+CreateShardSplitSubscriptions(List *targetNodeConnectionList,
+							  List *shardSplitSubscriberMetadataList,
+							  WorkerNode *sourceWorkerNode,
+							  char *superUser,
+							  char *databaseName)
+{
+	MultiConnection *targetConnection = NULL;
+	ShardSplitSubscriberMetadata *shardSplitPubSubMetadata = NULL;
+	forboth_ptr(targetConnection, targetNodeConnectionList,
+				shardSplitPubSubMetadata, shardSplitSubscriberMetadataList)
+	{
+		uint32 publicationForNodeId = shardSplitPubSubMetadata->slotInfo->targetNodeId;
+		Oid ownerId = shardSplitPubSubMetadata->tableOwnerId;
+		CreateShardSubscription(targetConnection,
+								sourceWorkerNode->workerName,
+								sourceWorkerNode->workerPort,
+								superUser,
+								databaseName,
+								ShardSplitPublicationName(publicationForNodeId, ownerId),
+								shardSplitPubSubMetadata->slotInfo->slotName,
+								ownerId);
+	}
+}
+
+
+/*
+ * WaitForShardSplitRelationSubscriptionsBecomeReady waits for a list of subscriptions
+ * to be come ready. This method invokes 'WaitForRelationSubscriptionsBecomeReady' for each
+ * subscription.
+ */
+void
+WaitForShardSplitRelationSubscriptionsBecomeReady(List *shardSplitPubSubMetadataList)
+{
+	ShardSplitSubscriberMetadata *shardSplitPubSubMetadata = NULL;
+	foreach_ptr(shardSplitPubSubMetadata, shardSplitPubSubMetadataList)
+	{
+		Bitmapset *tableOwnerIds = NULL;
+		tableOwnerIds = bms_add_member(tableOwnerIds,
+									   shardSplitPubSubMetadata->tableOwnerId);
+		WaitForRelationSubscriptionsBecomeReady(
+			shardSplitPubSubMetadata->targetNodeConnection, tableOwnerIds,
+			SHARD_SPLIT_SUBSCRIPTION_PREFIX);
+	}
+}
+
+
+/*
+ * WaitForShardSplitRelationSubscriptionsToBeCaughtUp waits until subscriptions are caught up till
+ * the source LSN. This method invokes 'WaitForShardSubscriptionToCatchUp' for each subscription.
+ */
+void
+WaitForShardSplitRelationSubscriptionsToBeCaughtUp(XLogRecPtr sourcePosition,
+												   List *shardSplitPubSubMetadataList)
+{
+	ShardSplitSubscriberMetadata *shardSplitPubSubMetadata = NULL;
+	foreach_ptr(shardSplitPubSubMetadata, shardSplitPubSubMetadataList)
+	{
+		Bitmapset *tableOwnerIds = NULL;
+		tableOwnerIds = bms_add_member(tableOwnerIds,
+									   shardSplitPubSubMetadata->tableOwnerId);
+
+		WaitForShardSubscriptionToCatchUp(shardSplitPubSubMetadata->targetNodeConnection,
+										  sourcePosition,
+										  tableOwnerIds,
+										  SHARD_SPLIT_SUBSCRIPTION_PREFIX);
+	}
+}
+
+
+char *
+DropExistingIfAnyAndCreateTemplateReplicationSlot(ShardInterval *shardIntervalToSplit,
+												  MultiConnection *sourceConnection)
+{
+	/*
+	 * To ensure SPLIT is idempotent drop any existing slot from
+	 * previous failed operation.
+	 */
+	StringInfo dropReplicationSlotCommand = makeStringInfo();
+	appendStringInfo(dropReplicationSlotCommand, "SELECT pg_drop_replication_slot('%s')",
```

Comment:
we already drop the replication slot in DropAllShardSplitLeftOvers, do we need this?

Reply:
ping

Reply:
@marcocitus: Yes we already drop it in DropAllShardSplitLeftOvers. We don't need this now.